ref(backend): Add AppStore Connect Credentials itunesCreated and itunesSessionRefreshAt properties

Swatinem opened the pull request:
This moves some code around, and saves the `refreshDate` of AppStore credentials and returns an approximated `expirationDate` on validation.

This is extracted from #26105 to at least land some of it.

Swatinem committed:
ref: Save AppStore Connect Credentials refreshDate

This moves some code around, and saves the `refreshDate` of AppStore credentials and returns an approximated `expirationDate` on validation.

priscilawebdev commented on static/app/components/modals/debugFileCustomRepository/appStoreConnect/index.tsx:
```diff
@@ -218,9 +218,9 @@ function AppStoreConnect({
+            defaultExpanded={!!appStoreConnectContext?.itunesSessionValid}
           >
-            {!!appStoreConnenctContext?.itunesSessionValid && (
+            {!!appStoreConnectContext?.itunesSessionValid && (
```
thanks for fixing this typo :)

priscilawebdev reviewed (dismissed):
lgtm

priscilawebdev commented on static/app/views/settings/project/appStoreConnectContext.tsx:
```diff
@@ -29,7 +29,7 @@ const Provider = withApi(
     function getAppStoreConnectSymbolSourceId() {
       return (project.symbolSources ? JSON.parse(project.symbolSources) : []).find(
-        symbolSouce => symbolSouce.type === 'appStoreConnect'
+        symbolSource => symbolSource.type === 'appStoreConnect'
```
and this too

RaduW commented on src/sentry/api/endpoints/project_app_store_connect_credentials.py:
```diff
@@ -151,6 +131,7 @@ def post(self, request, project):
             }
             credentials["encrypted"] = encrypt.encrypt_object(encrypted, key)
             credentials["type"] = "appStoreConnect"
+            credentials["refreshDate"] = datetime.utcnow()
             credentials["id"] = uuid4().hex
```
Updating the refresh date here is probably pretty safe (although technically incorrect).
We should remember when we got a success message from the 2fa (that's when the session was created).

I would still store the date in the session at the end of the 2fa call and retrieve it from there, saving the date at the end of 2fa is more important in the update call thought (see comment there).

RaduW commented on src/sentry/api/endpoints/project_app_store_connect_credentials.py:
```diff
@@ -238,6 +219,7 @@ def post(self, request, project, credentials_id):
             credentials.update(new_credentials)
 
             credentials["encrypted"] = encrypt.encrypt_object(secrets, key)
+            credentials["refreshDate"] = datetime.utcnow()
             credentials["id"] = uuid4().hex
```
Not a good idea to update the "refreshDate" here. A user might change the application id or ,  the appConnect info... all of these are not going to touch the itunes session, which is still going to expire after two weeks since it was obtained. 

A safer way to do it is to put the refreshDate into the sessionContext (at the end of 2fa) and then here if you can find a refreshDate it means that you just did a 2fa and you can refresh the date, if you don't find a "refreshDate" in the sessionContext that you shouldn't update it.

RaduW commented on src/sentry/utils/appleconnect/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        raise ValueError("bad sources") from e
+
+
+def validate_credentials(
+    project: Project, credentials_id: Optional[str]
+) -> Optional[AppConnectCredentialValidity]:
+    credentials = get_app_store_credentials(project, credentials_id)
+    key = project.get_option(credentials_key_name())
+
+    if key is None or credentials is None:
+        return None
+
+    expiration_date = None
+    if credentials.get("refreshDate") is not None:
+        expiration_date = (
+            datetime.datetime.fromisoformat(credentials.get("refreshDate")) + TOKEN_VALIDITY
+        )
+
+    secrets = encrypt.decrypt_object(credentials.get("encrypted"), key)
+
+    credentials = appstore_connect.AppConnectCredentials(
+        key_id=credentials.get("appconnectKey"),
+        key=secrets.get("appconnectPrivateKey"),
+        issuer_id=credentials.get("appconnectIssuer"),
+    )
+
+    session = requests.Session()
+    apps = appstore_connect.get_apps(session, credentials)
+
+    appstore_valid = apps is not None
+    itunes_connect.load_session_cookie(session, secrets.get("itunesSession"))
+    itunes_session_info = itunes_connect.get_session_info(session)
+
+    itunes_session_valid = itunes_session_info is not None
+
+    return AppConnectCredentialValidity(
+        appstore_credentials_valid=appstore_valid,
+        itunes_session_valid=itunes_session_valid,
+        expiration_date=expiration_date,
+    )
```
Don't really like this function :-) . 
It spans IMO to many concerns, I deals with HTTP sessions, encrypted content, project options, itunes_connect functionality.
I would keep the details of where I store credentials and how (encryption) at a separate level ( that's just my preference, it's not that how you do it is wrong).

RaduW commented on src/sentry/utils/appleconnect/__init__.py:
```diff
@@ -0,0 +1,86 @@
+)
+
+
+def credentials_key_name():
+    return "sentry:appleconnect_key"
+
+
+def symbol_sources_prop_name():
+    return "sentry:symbol_sources"
```
I wouldn't put it here ( in my mind at least) this is a lower level util that doesn't really deal with sentry storage (prop names... ).

RaduW reviewed (changes requested):
In general looks ok ( I personally wouldn't add encryption and storage concerns into `utils.appconnect` but that is just a preference).

I think we should change the way we save the session refresh date.
We should only change the refresh date when the user did a successful login (i.e. at the end of a 2fa) not whenever the user modifies the source (which might be modifying the appconnect private_key, or id , or app_id... ).

I've added a suggestion in the commentary where I think we should save the date

Commit:
feat: Introduce appstore connect functionality

This is a feature branch for appstore connect.

Commit:
ref(appconnect): Do not hide secrets

This makes sure that the itunes password and the appstore connect API
keys are returned together with the encrypted field.  This means we
will save them in the project details as symbolSource but we already
do this for other sources.

The reason to do this is that it allows the UI to load this and when a
user goes to modify the settings can pre-fill these, which simplifies
the UI and matches how S3 and GCS sources are handled.

priscilawebdev commented on src/sentry/utils/appleconnect/__init__.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def get_app_store_credentials(
+    project: Project, credentials_id: Optional[str]
```
should `credentials_id` be only of type string? It's a required parameter

priscilawebdev commented on src/sentry/utils/appleconnect/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        for source in sources:
+            if (
+                source.get("type") == "appStoreConnect"
+                and source.get("id") == credentials_id.lower()
```
shouldn't the `credential_id` be exactly the same as "id"?

flub commented:
apologies, i force-pushed this to fix up commits after i changed the target to the appstore connect feature branch

flub commented on src/sentry/utils/appleconnect/__init__.py:
```diff
@@ -0,0 +1,86 @@
+    expiration_date = None
+    if credentials.get("refreshDate") is not None:
+        expiration_date = (
+            datetime.datetime.fromisoformat(credentials.get("refreshDate")) + TOKEN_VALIDITY
```
This isn't supported on python 3.6.  You'll need something like `datetime.datetime.strptime(credentials.get("refershDate"), '%Y-%m-%d')`

flub replied:
apparently you also have time with sub-second precision.  this removes the sub-second precision:
`datetime.datetime.strptime(credentials.get("refreshDate").split('.')[0], '%Y-%m-%dT%H:%M:%S')`

flub replied:
or https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.isoparse apparently.  seems to already be in the dependencies

RaduW replied:
yes we already have dateutil so in order to support all formats for isodate ( with or without time, millisecs... )
we should use:
```
    from dateutil import parser
    # ...
    my_date = parser.isoparse("2020-10-22")
```

priscilawebdev committed:
Merge branch 'master' into feat/appstore-connect

priscilawebdev committed:
Merge branch 'master' into feat/appstore-connect

Commit:
ref(appstoreconnect): Add doc comments describing how this works (#26328)

Previously this was only on a notion page, makes more sense to have
the overview here.

Also change constants to be more python-style.

Commit:
Fix the schema that the ui now stores

Commit:
Merge branch 'feat/appstore-connect' into ref/appstore-saved

Commit:
Merge branch 'master' into feat/appstore-connect

Commit:
Merge branch 'feat/appstore-connect' into ref/appstore-saved

priscilawebdev committed:
Merge branch 'master' into feat/appstore-connect

Commit:
Move some logic back from sentry.utils

It contained logic about how to access project data etc, this is not
stuff that should be in sentry.utils which is only for interacting
with other things not for containing sentry knowledge.

Commit:
rename

Commit:
Set refresh date when it is refreshed

Also rename it to "itunesCreated" as that's more obvious.  Clarify and
rename the public API to this as well.

Fix the ISO date parsing by using the dateutil module.

Commit:
Need to validate some stuff.  but not now

Commit:
Ah, datetime...

Commit:
Merge branch 'master' into feat/appstore-connect

Commit:
Merge branch 'feat/appstore-connect' into ref/appstore-saved

Commit:
oh datetime *2

priscilawebdev reviewed (approved):
I tested it locally and everything seems to work

priscilawebdev committed:
Merge branch 'master' into feat/appstore-connect

priscilawebdev committed:
Merge branch 'feat/appstore-connect' into ref/appstore-saved

priscilawebdev commented on src/sentry/lang/native/symbolicator.py:
```diff
@@ -62,6 +62,9 @@
         "orgName": {"type": "string", "minLength": 1, "maxLength": 512},
         "encrypted": {"type": "string"},
+        "itunesCreated": {"type": "string"},
+        "itunesPassword": {"type": "string"},
+        "appconnectPrivateKey": {"type": "string"},
```
@flub will remove the `encrypted` prop this week

flub replied:
Yeah, they already exist here now (https://github.com/getsentry/sentry/blob/0ed459cc4cca0504f9de2cf80bbf8a3d7fb9fd21/src/sentry/api/endpoints/project_app_store_connect_credentials.py#L256-L257) and encrypted will be removed entirely because it serves no purpose anymore

priscilawebdev commented on static/app/types/debugFiles.tsx:
```diff
@@ -37,4 +37,6 @@ export type AppStoreConnectValidationData = {
   appstoreCredentialsValid: boolean;
   itunesSessionValid: boolean;
+  /** Approximate expiration Date of AppStore Credentials as an ISO DateTime */
+  expirationDate?: string;
```
yes in another PR

Commit:
Merge branch 'master' into ref/appstore-saved

RaduW reviewed (approved):
Date update looks ok now.
I think it's good to go

Commit:
Merge branch 'master' into ref/appstore-saved
